src/lib/traces/tracing-processor.js

```diff
@@ -97,21 +97,122 @@ class TraceProcessor {
-    const valueList = new traceviewer.metrics.ValueList();
-    traceviewer.metrics.sh.hazardMetric(valueList, model);
-    // grab last item, as it matches the fake idle we push()'d into the model'
-    const metricValue = valueList.valueDicts[valueList.valueDicts.length - 1];
-    return metricValue;
+  /**
+   * Find a main thread from supplied model with matching processId and
+   * threadId.
+   * @param {!Object} model TraceProcessor Model
+   * @param {number} processId
+   * @param {number} threadId
+   * @return {!Object}
+   * @private
+   */
+  static _findMainThreadFromIds(model, processId, threadId) {
+    const modelHelper = model.getOrCreateHelper(traceviewer.model.helpers.ChromeModelHelper);
+    const renderHelpers = traceviewer.b.dictionaryValues(modelHelper.rendererHelpers);
+    const mainThread = renderHelpers.find(helper => {
+      return helper.mainThread &&
+        helper.pid === processId &&
+        helper.mainThread.tid === threadId;
+    }).mainThread;
+
+    return mainThread;
+  }
+
+  /**
+   * Calculate duration at specified percentiles for given population of
+   * durations.
+   * @param {!Array<number>} durations
+   * @param {number} totalTime
+   * @param {!Array<number>} percentiles
+   * @return {!Array<{percentile: number, time: number}>}
+   * @private
+   */
+  static _riskPercentiles(durations, totalTime, percentiles) {
+    durations.sort((a, b) => a - b);
+    let busyTime = 0;
+    for (let i = 0; i < durations.length; i++) {
+      busyTime += durations[i];
+    }
+
+    // Start with idle time already complete.
+    let completedTime = totalTime - busyTime;
+    let duration = 0;
+    let cdfTime = completedTime;
+    let remainingCount = durations.length + 1;
+    const results = [];
+
+    // Find percentiles of interest, in order.
+    for (let percentile of percentiles) {
+      // Loop over durations, calculating a CDF value for each until it is above
+      // the target percentile.
+      const percentileTime = percentile * totalTime;
+      while (cdfTime < percentileTime && remainingCount > 1) {
+        completedTime += duration;
+        remainingCount--;
+        duration = durations[durations.length - remainingCount];
+
+        // Calculate value of CDF (multiplied by totalTime) for the end of this duration.
+        cdfTime = completedTime + duration * remainingCount;
+      }
+
+      // Negative results are within idle time (0ms wait by definition), so clamp at zero.
+      results.push({
+        percentile,
+        time: Math.max(0, (percentileTime - completedTime) / remainingCount)
+      });
+    }
+
+    return results;
+  }
+
+  /**
+   * Calculates the maximum queueing time (in ms) of high priority tasks for
+   * selected percentiles within a window of the main thread.
+   * @param {!Array<!Object>} trace
+   * @param {number=} startTime Optional start time of range of interest. Defaults to trace start.
```

Comment:
do me a favor and denote what sort of timestamp is used?  high-res time with milliseconds or nanoseconds?